File: src/test/java/com/github/njuro/jard/board/BoardIntegrationTest.java
package com.github.njuro.jard.board;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.github.njuro.jard.attachment.AttachmentCategory;
import com.github.njuro.jard.common.Constants;
import com.github.njuro.jard.common.Mappings;
import com.github.njuro.jard.common.MockRequestTest;
import com.github.njuro.jard.common.WithMockUserAuthorities;
import com.github.njuro.jard.user.UserAuthority;
import java.util.List;
import java.util.Set;
import org.assertj.core.util.Sets;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpMethod;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest
@Transactional
class BoardIntegrationTest extends MockRequestTest {

  private static final String API_ROOT = Mappings.API_ROOT_BOARDS;

  @Autowired private BoardFacade boardFacade;

  private BoardForm boardForm;

  @BeforeEach
  public void setUp() {
    boardForm =
        BoardForm.builder()
            .name("Random")
            .label("r")
            .boardSettingsForm(
                BoardSettingsForm.builder()
                    .bumpLimit(Constants.MAX_BUMP_LIMIT)
                    .threadLimit(Constants.MAX_THREAD_LIMIT)
                    .attachmentCategories(Sets.newLinkedHashSet(AttachmentCategory.values()))
                    .nsfw(true)
                    .defaultPosterName("Anonymous")
                    .forceDefaultPosterName(true)
                    .build())
            .build();
  }

  @Test
  @WithMockUserAuthorities(UserAuthority.MANAGE_BOARDS)
  void testCreateBoard() throws Exception {
    performMockRequest(HttpMethod.PUT, API_ROOT, boardForm)
        .andExpect(status().isOk())
        .andExpect(content().json(toJson(boardForm.toBoard())));

    assertThat(boardFacade.resolveBoard(boardForm.getLabel())).isNotNull();
  }

  @Test
  @WithMockUserAuthorities(UserAuthority.MANAGE_BOARDS)
  void testGetAttachmentCategories() throws Exception {
    var result =
        performMockRequest(HttpMethod.GET, API_ROOT + "/attachment-categories", boardForm)
            .andExpect(status().isOk())
            .andExpect(nonEmptyBody())
            .andReturn();

    assertThat(getResponseCollection(result, Set.class, AttachmentCategory.Preview.class))
        .containsExactlyInAnyOrderElementsOf(boardFacade.getAttachmentCategories());
  }

  @Test
  void testGetAllBoards() throws Exception {
    boardFacade.createBoard(boardForm);

    var result =
        performMockRequest(HttpMethod.GET, API_ROOT)
            .andExpect(status().isOk())
            .andExpect(nonEmptyBody())
            .andReturn();

    assertThat(getResponseCollection(result, List.class, Board.class))
        .containsExactlyInAnyOrderElementsOf(boardFacade.getAllBoards());
  }

  @Test
  void testGetBoard() throws Exception {
    boardFacade.createBoard(boardForm);

    var result =
        performMockRequest(
                HttpMethod.GET,
                buildUri(API_ROOT + Mappings.PATH_VARIABLE_BOARD, boardForm.getLabel()))
            .andExpect(status().isOk())
            .andExpect(nonEmptyBody())
            .andReturn();

    assertThat(getResponse(result, Board.class))
        .extracting(Board::getLabel)
        .isEqualTo(boardForm.getLabel());
  }

  @Test
  void testGetBoardCatalog() throws Exception {
    boardFacade.createBoard(boardForm);

    var result =
        performMockRequest(
                HttpMethod.GET,
                buildUri(
                    API_ROOT + Mappings.PATH_VARIABLE_BOARD + "/catalog", boardForm.getLabel()))
            .andExpect(status().isOk())
            .andExpect(nonEmptyBody())
            .andReturn();

    assertThat(getResponse(result, Board.class))
        .extracting(Board::getLabel)
        .isEqualTo(boardForm.getLabel());
  }

  @Test
  @WithMockUserAuthorities(UserAuthority.MANAGE_BOARDS)
  void testEditBoard() throws Exception {
    boardFacade.createBoard(boardForm);

    assertThat(boardFacade.resolveBoard(boardForm.getLabel()))
        .isNotNull()
        .extracting(Board::getName)
        .isEqualTo(boardForm.getName());

    String newName = "Technology";
    boardForm.setName(newName);
    performMockRequest(
            HttpMethod.POST,
            buildUri(API_ROOT + Mappings.PATH_VARIABLE_BOARD + "/edit", boardForm.getLabel()),
            boardForm)
        .andExpect(status().isOk());

    assertThat(boardFacade.resolveBoard(boardForm.getLabel()))
        .isNotNull()
        .extracting(Board::getName)
        .isEqualTo(newName);
  }

  @Test
  @WithMockUserAuthorities(UserAuthority.MANAGE_BOARDS)
  void testDeleteBoard() throws Exception {
    boardFacade.createBoard(boardForm);

    assertThat(boardFacade.resolveBoard(boardForm.getLabel())).isNotNull();

    performMockRequest(
            HttpMethod.DELETE,
            buildUri(API_ROOT + Mappings.PATH_VARIABLE_BOARD, boardForm.getLabel()))
        .andExpect(status().isOk());

    assertThatThrownBy(() -> boardFacade.resolveBoard(boardForm.getLabel()))
        .isInstanceOf(BoardNotFoundException.class);
  }
}

File: src/test/kotlin/com/github/njuro/jard/board/BoardIntegrationTest.kt
package com.github.njuro.jard.board

import com.github.njuro.jard.MockMvcTest
import com.github.njuro.jard.TestDataRepository
import com.github.njuro.jard.WithContainerDatabase
import com.github.njuro.jard.WithMockJardUser
import com.github.njuro.jard.attachment.AttachmentCategory
import com.github.njuro.jard.board
import com.github.njuro.jard.board.dto.BoardDto
import com.github.njuro.jard.board.dto.BoardForm
import com.github.njuro.jard.common.InputConstraints.MAX_BOARD_NAME_LENGTH
import com.github.njuro.jard.common.Mappings
import com.github.njuro.jard.randomString
import com.github.njuro.jard.thread
import com.github.njuro.jard.toForm
import com.github.njuro.jard.user.UserAuthority
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.collections.shouldNotBeEmpty
import io.kotest.matchers.optional.shouldBeEmpty
import io.kotest.matchers.optional.shouldBePresent
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.web.servlet.delete
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import org.springframework.test.web.servlet.put
import org.springframework.transaction.annotation.Transactional

@WithContainerDatabase
@Transactional
internal class BoardIntegrationTest : MockMvcTest() {

    @Autowired
    private lateinit var db: TestDataRepository

    @Nested
    @DisplayName("create board")
    @WithMockJardUser(UserAuthority.MANAGE_BOARDS)
    inner class CreateBoard {
        private fun createBoard(boardForm: BoardForm) =
            mockMvc.post(Mappings.API_ROOT_BOARDS) { setUp(); body(boardForm) }

        @Test
        fun `create valid board`() {
            val boardForm = board(label = "r").toForm()

            val response = createBoard(boardForm).andExpect { status { isCreated() } }.andReturnConverted<BoardDto>()
            response.label shouldBe boardForm.label
        }

        @Test
        fun `don't create invalid board`() {
            val boardForm = board(label = "r", name = randomString(MAX_BOARD_NAME_LENGTH + 1)).toForm()

            createBoard(boardForm).andExpect { status { isBadRequest() } }
        }
    }

    @Test
    @WithMockJardUser(UserAuthority.MANAGE_BOARDS)
    fun `get attachment categories`() {
        mockMvc.get("${Mappings.API_ROOT_BOARDS}/attachment-categories") { setUp() }.andExpect { status { isOk() } }
            .andReturnConverted<Set<AttachmentCategory.Preview>>().shouldNotBeEmpty()
    }

    @Test
    fun `get all boards`() {
        db.insert(board(label = "r"))
        db.insert(board(label = "fit"))
        db.insert(board(label = "sp"))
        mockMvc.get(Mappings.API_ROOT_BOARDS) { setUp() }.andExpect { status { isOk() } }
            .andReturnConverted<List<BoardDto>>() shouldHaveSize 3
    }

    @Nested
    @DisplayName("get board")
    inner class GetBoard {
        private fun getBoard(label: String) = mockMvc.get("${Mappings.API_ROOT_BOARDS}/$label") { setUp() }

        @Test
        fun `get existing board`() {
            val board = db.insert(board(label = "r"))

            getBoard(board.label).andExpect { status { isOk() } }
                .andReturnConverted<BoardDto>().label shouldBe board.label
        }

        @Test
        fun `don't get non-existing board`() {
            getBoard("xxx").andExpect { status { isNotFound() } }
        }
    }

    @Nested
    @DisplayName("get board catalog")
    inner class GetBoardCatalog {
        private fun getBoardCatalog(label: String) =
            mockMvc.get("${Mappings.API_ROOT_BOARDS}/$label/catalog") { setUp() }

        @Test
        @Suppress("UNUSED_VARIABLE")
        fun `get existing board catalog`() {
            val board = db.insert(board(label = "r"))
            val thread = db.insert(thread(board))

            val response = getBoardCatalog(board.label).andExpect { status { isOk() } }.andReturnConverted<BoardDto>()
            response.label shouldBe board.label
            response.threads shouldHaveSize 1
        }

        @Test
        fun `don't get non-existing board`() {
            getBoardCatalog("xxx").andExpect { status { isNotFound() } }
        }
    }

    @Nested
    @DisplayName("edit board")
    @WithMockJardUser(UserAuthority.MANAGE_BOARDS)
    inner class EditBoard {
        private fun editBoard(boardForm: BoardForm) =
            mockMvc.put("${Mappings.API_ROOT_BOARDS}/${boardForm.label}") { setUp(); body(boardForm) }

        @Test
        fun `edit valid board`() {
            val board = db.insert(board(label = "r", name = "random"))
            val updatedName = "Updated"

            val response = editBoard(board.apply { name = updatedName }.toForm()).andExpect { status { isOk() } }
                .andReturnConverted<BoardDto>()
            response.name shouldBe updatedName
            db.select(board).shouldBePresent { it.name shouldBe updatedName }
        }

        @Test
        fun `don't edit non-existing board`() {
            editBoard(board(label = "r", name = "random").toForm()).andExpect { status { isNotFound() } }
        }

        @Test
        fun `don't edit invalid board`() {
            val board = db.insert(board(label = "r", name = "random"))

            editBoard(
                board(label = board.label, name = randomString(MAX_BOARD_NAME_LENGTH + 1))
                    .toForm()
            ).andExpect { status { isBadRequest() } }
            db.select(board).shouldBePresent { it.name shouldBe board.name }
        }
    }

    @Nested
    @DisplayName("delete board")
    @WithMockJardUser(UserAuthority.MANAGE_BOARDS)
    inner class DeleteBoard {
        private fun deleteBoard(label: String) = mockMvc.delete("${Mappings.API_ROOT_BOARDS}/$label") { setUp() }

        @Test
        fun `delete existing board`() {
            val board = db.insert(board(label = "r"))

            deleteBoard(board.label).andExpect { status { isOk() } }
            db.select(board).shouldBeEmpty()
        }
    }
}
